<h1>Сервис для чтения книг<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-и-задачи:" data-toc-modified-id="Цели-и-задачи:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели и задачи:</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#books" data-toc-modified-id="books-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>books</a></span></li><li><span><a href="#authors" data-toc-modified-id="authors-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>authors</a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>publishers</a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>ratings</a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>reviews</a></span></li></ul></li><li><span><a href="#Запросы" data-toc-modified-id="Запросы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Запросы</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li></ul></div>

# Сервис для чтения книг

## Цели и задачи:

Компания купила крупный сервис для чтения книг по подписке. Задача  — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

Эти данные помогут сформулировать ценностное предложение для нового продукта.

Наши задания: 

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных

**Таблица books**

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## Импорт библиотек

In [14]:
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

CSS = """ 
.output { 
    flex-direction: row; 
} 
""" 

In [15]:
def table_info(table, table_name):
    
    #table = data[table]
    display(f"Таблица {table_name}")
    display(f"Количество строк: {table.shape[0]}, Количество столбцов: {table.shape[1]}")
    for column in table.columns:
        display(f"Столбец: {column}", 
                f"Количество уникальных значений: {table[column].nunique()}",
                f"Тип данных: {table[column].dtype}")
    display(f"Количество явных дубликатов: {table.duplicated().sum()}")
    display(pd.DataFrame(round(table.isna().mean()*100, 2)).style.background_gradient('coolwarm'))
    display(table.head())

In [16]:
# параметры подключения к БД
db_config = {
    'user': 'praktikum_student', 
    'pwd': 'Sdf4$2;d-d30pp', 
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db' # название базы данных
}

# создаем строку подключения к БД
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

# создаем объект engine
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# список таблиц
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

# создаем словарь для хранения таблиц
data = {}

# выводим данные из каждой таблицы
for table in tables:
    query = f"SELECT * FROM {table}"
    df = pd.read_sql(query, con=engine)
    data[table] = df



## Исследование таблиц

### books

In [17]:
books = data['books']

table_info(books, 'books')

'Таблица books'

'Количество строк: 1000, Количество столбцов: 6'

'Столбец: book_id'

'Количество уникальных значений: 1000'

'Тип данных: int64'

'Столбец: author_id'

'Количество уникальных значений: 636'

'Тип данных: int64'

'Столбец: title'

'Количество уникальных значений: 999'

'Тип данных: object'

'Столбец: num_pages'

'Количество уникальных значений: 454'

'Тип данных: int64'

'Столбец: publication_date'

'Количество уникальных значений: 618'

'Тип данных: object'

'Столбец: publisher_id'

'Количество уникальных значений: 340'

'Тип данных: int64'

'Количество явных дубликатов: 0'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Пропуски дубликаты в данных не обнаружены. Тип данных в столбце с датой изменен.

### authors

In [18]:
authors = data['authors']

table_info(authors, "authors")

'Таблица authors'

'Количество строк: 636, Количество столбцов: 2'

'Столбец: author_id'

'Количество уникальных значений: 636'

'Тип данных: int64'

'Столбец: author'

'Количество уникальных значений: 636'

'Тип данных: object'

'Количество явных дубликатов: 0'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Пропуски дубликаты в данных не обнаружены.

### publishers

In [19]:
publishers = data['publishers']

table_info(publishers, "publishers")

'Таблица publishers'

'Количество строк: 340, Количество столбцов: 2'

'Столбец: publisher_id'

'Количество уникальных значений: 340'

'Тип данных: int64'

'Столбец: publisher'

'Количество уникальных значений: 340'

'Тип данных: object'

'Количество явных дубликатов: 0'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Пропуски дубликаты в данных не обнаружены.

### ratings

In [20]:
ratings = data['ratings']

table_info(ratings, "ratings")

'Таблица ratings'

'Количество строк: 6456, Количество столбцов: 4'

'Столбец: rating_id'

'Количество уникальных значений: 6456'

'Тип данных: int64'

'Столбец: book_id'

'Количество уникальных значений: 1000'

'Тип данных: int64'

'Столбец: username'

'Количество уникальных значений: 160'

'Тип данных: object'

'Столбец: rating'

'Количество уникальных значений: 5'

'Тип данных: int64'

'Количество явных дубликатов: 0'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Пропуски дубликаты в данных не обнаружены.

### reviews

In [21]:
reviews = data['reviews']

table_info(reviews, "reviews")

'Таблица reviews'

'Количество строк: 2793, Количество столбцов: 4'

'Столбец: review_id'

'Количество уникальных значений: 2793'

'Тип данных: int64'

'Столбец: book_id'

'Количество уникальных значений: 994'

'Тип данных: int64'

'Столбец: username'

'Количество уникальных значений: 160'

'Тип данных: object'

'Столбец: text'

'Количество уникальных значений: 2793'

'Тип данных: object'

'Количество явных дубликатов: 0'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

В нашем распоряжении база данных состоящая из пяти таблиц:

- books
- authors
- publishers
- ratings
- reviews

Пропуски дубликаты в таблицах не обнаружены.

## Запросы

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [22]:
query_1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

req_1 = pd.io.sql.read_sql(query_1, con = engine)
req_1
display(f'После 1 января 2000 года вышло {req_1.iloc[0,0]} книг')

'После 1 января 2000 года вышло 819 книг'

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [23]:
query_2 = ''' 

SELECT 
    books.title AS book_name,
    sub_rev.n_rev AS reviews,
    sub_rat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS sub_rat ON sub_rat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS sub_rev ON sub_rev.book_id = books.book_id
ORDER BY avg_rating DESC               
; ''' 

req_2 = pd.io.sql.read_sql(query_2, con = engine)
req_2

book_name  reviews  avg_rating
0                 Pop Goes the Weasel (Alex Cross  #5)      2.0        5.00
1    The Ghost Map: The Story of London's Most Terr...      2.0        5.00
2    In the Hand of the Goddess (Song of the Liones...      2.0        5.00
3                             Tai-Pan (Asian Saga  #2)      2.0        5.00
4    How to Be a Domestic Goddess: Baking and the A...      1.0        5.00
..                                                 ...      ...         ...
995  The World Is Flat: A Brief History of the Twen...      3.0        2.25
996                                              Junky      2.0        2.00
997                  His Excellency: George Washington      2.0        2.00
998                                      Drowning Ruth      3.0        2.00
999                               Harvesting the Heart      2.0        1.50

[1000 rows x 3 columns]

Для каждой книги посчитано количество обзоров и средняя оценка.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [11]:
query_3 = ''' 
SELECT
    P.publisher_id,
    P.publisher,
    P.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS P
LIMIT 1
; ''' 

req_3 = pd.io.sql.read_sql(query_3, con = engine)
req_3

publisher_id      publisher  n_books
0           212  Penguin Books       42

Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query_4 = ''' 
SELECT 
    aut_avg.author,
    AVG(aut_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            authors.author AS author,
            AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
             books.book_id,
             authors.author_id
        HAVING
            COUNT(ratings.rating_id) >=50 
        ) AS aut_avg 
GROUP BY    
    aut_avg.author
ORDER BY
    avg_avg_rating DESC
LIMIT 5
    
; ''' 

req_4 = pd.io.sql.read_sql(query_4, con = engine)
req_4

author  avg_avg_rating
0         J.K. Rowling/Mary GrandPré        4.283844
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.258446
3                  Louisa May Alcott        4.192308
4                       Rick Riordan        4.080645

Автор с самой высокой средней оценков книг - J.K. Rowling/Mary GrandPré (учитаны только книги с 50 и более оценками).

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
query_5 = ''' 

SELECT
    AVG(Sub.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS Sub
; ''' 

req_5 = pd.io.sql.read_sql(query_5, con = engine)
req_5

avg
0  24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33. 

# Выводы:

При анализе базы данных купленного сервиса для чтения книг, были выявлено следующее :

- Пропуски дубликаты в таблицах не обнаружены.

- После 1 января 2000 года вышло 819 книг.

- Для каждой книги посчитано количество обзоров и средняя оценка:

- - 0   Pop Goes the Weasel (Alex Cross #5)		2.0	5.00

- - 1	The Ghost Map: The Story of London's Most Terr...		2.0	5.00

- - 2	In the Hand of the Goddess (Song of the Liones...		2.0	5.00

- - 3	Tai-Pan (Asian Saga #2)	610	2.0	5.00

- - 4	How to Be a Domestic Goddess: Baking and the A...		1.0	5.00

...	...	...	...	...

- - 995	The World Is Flat: A Brief History of the Twen...		3.0	2.25

- - 996	Junky		2.0	2.00

- - 997	His Excellency: George Washington		2.0	2.00

- - 998	Drowning Ruth		3.0	2.00

- - 999	Harvesting the Heart		2.0	1.50

- Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц - 42 книги.

- Автор с самой высокой средней оценков книг - J.K. Rowling/Mary GrandPré (учитаны только книги с 50 и более оценками).

- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33.